src/box/xrow_buf.h

```diff
@@ -0,0 +1,80 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2024, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#pragma once
+
+#include "trivia/config.h"
+
+#if defined(ENABLE_XROW_BUF)
+#include "xrow_buf_impl.h"
+#else /* !defined(ENABLE_XROW_BUF) */
+
+#include <sys/uio.h>
+#include "xrow.h"
+#include "small/region.h"
+
+typedef int
+(*xrow_buf_flush_cb)(void *cb_data, struct iovec *iov, int iovcnt);
+
+/** A buffer for encoded rows. */
+struct xrow_buf {
+	/** A callback invoked each time xrow_buf is flushed. */
+	xrow_buf_flush_cb flush_cb;
+	/** Data for \a flush_cb. */
+	void *flush_cb_data;
+};
```

Comment:
A callback which can implicitly on any push perform a flush looks like an overkill and makes relay code harder to understand what happens when exactly. I would suggest to keep `xrow_buf` a simple buffer. And make relay after each push check if the buffer needs to be flushed, and do the flush then.

Which in case of CE apparently will mean that the buffer can store only a single packet at once.